core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+pub struct ThreadAwareAccountLocks {
+    /// Number of threads.
+    num_threads: usize, // 0..MAX_THREADS
+    /// Limit on the number of sequentially-queued transactions per account.
+    sequential_queue_limit: u32,
+    /// Write locks - only on thread can hold a write lock at a time.
+    /// Contains how many write locks are held by the thread.
+    write_locks: HashMap<Pubkey, (ThreadId, u32)>,
+    /// Read locks - multiple threads can hold a read lock at a time.
+    /// Contains thread-set for easily checking which threads are scheduled.
+    /// Contains how many read locks are held by each thread.
+    read_locks: HashMap<Pubkey, (ThreadSet, [u32; MAX_THREADS])>,
+}
+
+impl ThreadAwareAccountLocks {
+    /// Creates a new `ThreadAwareAccountLocks` with the given number of threads
+    /// and queue limit.
+    pub fn new(num_threads: usize, sequential_queue_limit: u32) -> Self {
+        assert!(num_threads > 0, "num threads must be > 0");
+        assert!(
+            num_threads <= MAX_THREADS,
+            "num threads must be <= {MAX_THREADS}"
+        );
+        assert!(
+            sequential_queue_limit > 0,
+            "sequential queue limit must be > 0"
+        );
+
+        Self {
+            num_threads,
+            sequential_queue_limit,
+            write_locks: HashMap::new(),
+            read_locks: HashMap::new(),
+        }
+    }
+
+    /// Returns the `ThreadId` if the accounts are able to be locked
+    /// for the given thread, otherwise `None` is returned.
+    /// If accounts are schedulable, then they are locked for the thread
+    /// selected by the `thread_selector` function.
+    pub fn try_lock_accounts<'a>(
+        &mut self,
+        write_account_locks: impl Iterator<Item = &'a Pubkey> + Clone,
+        read_account_locks: impl Iterator<Item = &'a Pubkey> + Clone,
+        thread_selector: impl FnOnce(ThreadSet) -> ThreadId,
+    ) -> Option<ThreadId> {
+        let schedulable_threads = self
+            .accounts_schedulable_threads(write_account_locks.clone(), read_account_locks.clone());
+        (!schedulable_threads.is_empty()).then(|| {
+            let thread_id = thread_selector(schedulable_threads);
+            self.lock_accounts(write_account_locks, read_account_locks, thread_id);
+            thread_id
+        })
+    }
+
+    /// Returns `ThreadSet` that the given accounts can be scheduled on.
+    fn accounts_schedulable_threads<'a>(
+        &self,
+        write_account_locks: impl Iterator<Item = &'a Pubkey>,
+        read_account_locks: impl Iterator<Item = &'a Pubkey>,
+    ) -> ThreadSet {
+        let mut schedulable_threads = ThreadSet::any(self.num_threads);
+
+        for account in write_account_locks {
+            schedulable_threads &= self.write_schedulable_threads(account);
+        }
+
+        for account in read_account_locks {
+            schedulable_threads &= self.read_schedulable_threads(account);
```

Comment:
I doubt rustc is smart to enough to return early once `scheduleable_threads == ThreadSet::none()`.... So, make `{write,read}_schedulable_threads()` return `Option<_>` and do `?` here?

Reply:
that's a good point, however just doing ? is unfortunately not strong enough here. I will explicitly check the `schedulable` threads in each loop.

Example:
We could potentially have account 1 which is schedulable only on thread 1, and account 2 which is schedulable only on thread 2. Neither `{write|read}_schedulable_threads` returns an empty set, but we end up with one after the `&=`.

Reply:
was done here: 3c92bac18e5ed1b9dd4f04c5f5012a117ce6bca5